File: src/ros2_bridge/mappers/nav_msgs/occupancy_grid.rs
//! Typed mapper for `nav_msgs/msg/OccupancyGrid`.
//!
//! Owned `ros_to_bus` so occupancy `data` (`int8[]` ↔ proto `bytes`) moves
//! instead of element-wise copies.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn occupancy_grid_to_bus(
    msg: ros_env::nav_msgs::msg::OccupancyGrid,
) -> crate::nav_msgs::msg::v1::OccupancyGrid {
    crate::nav_msgs::msg::v1::OccupancyGrid {
        header: Some(crate::ros2_bridge::mappers::std_msgs::header::header_to_bus(msg.header)),
        info: Some(
            crate::ros2_bridge::mappers::nav_msgs::map_meta_data::map_meta_data_to_bus(msg.info),
        ),
        data: crate::ros2_bridge::mappers::convert::i8_seq_to_bytes(msg.data),
    }
}

pub(crate) fn occupancy_grid_to_ros(
    bus: crate::nav_msgs::msg::v1::OccupancyGrid,
) -> ros_env::nav_msgs::msg::OccupancyGrid {
    ros_env::nav_msgs::msg::OccupancyGrid {
        header: crate::ros2_bridge::mappers::std_msgs::header::header_to_ros(
            bus.header.unwrap_or_default(),
        ),
        info: crate::ros2_bridge::mappers::nav_msgs::map_meta_data::map_meta_data_to_ros(
            bus.info.unwrap_or_default(),
        ),
        data: crate::ros2_bridge::mappers::convert::bytes_to_i8_seq(bus.data),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct NavMsgsOccupancyGridMapper;

impl TypedTopicMapper for NavMsgsOccupancyGridMapper {
    type Ros = ros_env::nav_msgs::msg::OccupancyGrid;
    type Bus = crate::nav_msgs::msg::v1::OccupancyGrid;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(occupancy_grid_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(occupancy_grid_to_ros(msg))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn occupancy_grid_i8_bytes_roundtrip() {
        let ros = ros_env::nav_msgs::msg::OccupancyGrid {
            header: Default::default(),
            info: Default::default(),
            data: vec![0, 1, -1, 100],
        };
        let bus = occupancy_grid_to_bus(ros);
        assert_eq!(bus.data, vec![0, 1, 255, 100]);
        let back = occupancy_grid_to_ros(bus);
        assert_eq!(back.data, vec![0, 1, -1, 100]);
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/nav_msgs/occupancy_grid.py
"""Generated mapper for `nav_msgs/msg/OccupancyGrid`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.std_msgs.header import header_to_bus, header_to_ros
from robot_bus.ros2_bridge.mappers.nav_msgs.map_meta_data import map_meta_data_to_bus, map_meta_data_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.nav_msgs.msg.v1 import OccupancyGrid as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def occupancy_grid_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.header.CopyFrom(header_to_bus(msg.header))
    bus.info.CopyFrom(map_meta_data_to_bus(msg.info))
    bus.data = _convert.i8_seq_to_bytes(msg.data)
    return bus


def occupancy_grid_to_ros(bus):
    from nav_msgs.msg import OccupancyGrid as RosMsg

    out = RosMsg()
    out.header = header_to_ros(bus.header)
    out.info = map_meta_data_to_ros(bus.info)
    out.data = _convert.bytes_to_i8_seq(bus.data)
    return out


class NavMsgsOccupancyGridMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from nav_msgs.msg import OccupancyGrid as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return occupancy_grid_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return occupancy_grid_to_ros(bus)
